### PR TITLE
Add luci.dart, a felt replacement for LUCI

### PR DESCRIPTION
**Pull request by yjbanov**
## Background

We have been calling `felt` directly from the LUCI recipes. Unfortunately `felt` is not a self-contained build tool. It requires that LUCI recipes do ad hoc setup for every test target they run. All test targets are hard-coded in the recipes, which necessitates recipe changes every time we need to do something different.

## luci.dart

[Design doc](https://docs.google.com/document/d/14_Lbe3d7P73xfJvOsS5NPNF8L9VZOvV9Kn4Z2YzyUVw).

This PR includes defining targets, enumerating targets, and running targets.

This PR does _not_ include target dependencies and output caching.

This tool sits between LUCI and felt (and other commands we run as part of the build, such as `pub get`). LUCI recipes should only ever need to call `luci.dart` and not hard-code any other target-specific logic.

See the `README.md` for what `luci.dart` supports. Essentially, it allows enumerating all available targets, and running targets. When we switch from `felt` to `luci.dart` the recipes will no longer need to hard-code `felt check_licenses` and `felt test --browser=firefox` any more. Instead, LUCI can query for the available targets. Targets can be filtered based on the agent profile. For example, some targets require Linux, other require Mac. Each target comes with `agentProfiles` which lists agent types the target can run on. This allows the recipe to launch the appropriate agent type, rather than, again, assuming any one particular type.

### CODE REVIEW

File: tools/luci/lib/luci_common.dart [hunk 1]
```diff
@@ -0,0 +1,400 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply tools/luci/lib/luci_common.dart:1]: Probably because this file is some combination of `environment.dart`, `utils.dart`, and `exceptions.dart`.

[the remark this answers is not in the document]
Comment (nturgut) [reply tools/luci/lib/luci_common.dart:1]: In the last 6 months I put a lot of effort to split these files and keep every file small. Please split them otherwise I cannot approve this change. (Since I'm the person who works most using these files, and would probably would lose the most productivity over these huge files, I think my request being blocking makes sense :) )

File: tools/luci/lib/luci_targets.dart [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    name: '//e2etests/web:integration_tests_linux_chrome',
+    agentProfiles: kLinuxAgent,
+    runner: WebIntegrationTestsRunner(
+      browser: 'chrome',
+    ),
+    environment: <String, String>{
+      'CHROME_NO_SANDBOX': 'true',
+    },
+  ),
+
```
Comment (nturgut) [inline tools/luci/lib/luci_targets.dart:47]: Integration test runner is under lib/web_ui.

File: lib/web_ui/dev/test_runner.dart [hunk 3]
```diff
@@ -381,7 +379,7 @@ class TestCommand extends Command<bool> with ArgUtils {
     );
 
     if (exitCode != 0) {
-      throw ToolException(
```
Comment (nturgut) [inline lib/web_ui/dev/test_runner.dart:384]: Can you please remove the name change. Let's discuss name changes on these files (and also location refactors) before you start a PR.

**Comment by nturgut** [comment]
I liked the runner idea, and the targets.

I would prefer you submit only those parts and do not the the other changes (renaming, location changes.)

File: tools/luci/lib/luci_common.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply tools/luci/lib/luci_common.dart:1]: Done.

File: tools/luci/lib/luci_targets.dart (hunk 2 above)
Comment (yjbanov) [reply tools/luci/lib/luci_targets.dart:47]: The paths determine the locations of build artifacts (the artifact caching is not part of this PR, but I'm establishing the convention now while I'm introducing targets). We can use the same runner for multiple test targets, each in its own directory outputting its own files. Since integration tests live under e2etests, their outputs also live there.

This will allow us to compile on Linux but test on Windows.

File: lib/web_ui/dev/test_runner.dart (hunk 3 above)
Comment (yjbanov) [reply lib/web_ui/dev/test_runner.dart:384]: Done. I don't mind discussing code changes in the PR. The reason for rename was [consistency with flutter/flutter](https://github.com/flutter/flutter/blob/b24bca1383fa8553e5a39a831fb19ee214ef233f/packages/flutter_tools/lib/src/base/common.dart#L21), and because we don't use it as an exception (e.g. we don't care about the stack trace).

**Comment by yjbanov** [comment]
> I would prefer you submit only those parts and do not the the other changes (renaming, location changes.)

Undid the renaming and file split. I have to change the location for the following reasons:

- `felt` code lives under the `dev/` directory. Dart only supports importing from `lib/`.
- We also apply different license rules for tools vs code that's shipped with the SDK, so can't put it in `web_ui/lib`.
- `luci.dart` is a generic tool that can be used for non-Web targets, and therefore should not have a dependency on the `web_ui` directory.

In other words, a new location is necessary. I'm open to suggestions, but thought that `//tools/luci` made the most sense.

**Comment by nturgut** [comment]
> > I would prefer you submit only those parts and do not the the other changes (renaming, location changes.)
> 
> Undid the renaming and file split. I have to change the location for the following reasons:
> 
> * `felt` code lives under the `dev/` directory. Dart only supports importing from `lib/`.
> * We  apply different license rules for tools vs code that's shipped with the SDK, so can't put it in `web_ui/lib`.
> * `luci.dart` is a generic tool that can be used for non-Web targets, and therefore should not have a dependency on the `web_ui` directory.
> 
> In other words, a new location is necessary. I'm open to suggestions, but thought that `//tools/luci` made the most sense.

As far as I understand, this will only be used by our team for 1-2 quarters until infra team creates a general solution.  I think one rule of thumb (we discussed together before) is not to spend time on a feature until its use is clear/immediate. Since there are no immediate other clients to this tool, can we carry it to web_ui for now.

The advantage of having luci.dart under tools is not clear to me. On the other hand, I can use the time for rebasing many PRs for focusing on other issues.

File: .cirrus.yml [hunk 4]
```diff
@@ -88,11 +88,14 @@ task:
...
         $ENGINE_PATH/src/out/host_debug_unopt/dart-sdk/bin/pub get
-        export FELT="$ENGINE_PATH/src/out/host_debug_unopt/dart-sdk/bin/dart dev/felt.dart"
-        $FELT check-licenses
-        CHROME_NO_SANDBOX=true $FELT test
+        export LUCI="$ENGINE_PATH/src/out/host_debug_unopt/dart-sdk/bin/dart bin/luci.dart"
+        $LUCI targets
+        $LUCI targets --pretty
+        $LUCI run //lib/web_ui:check_licenses
+        $LUCI run //lib/web_ui:unit_tests_linux_chrome
+        $LUCI run //e2etests/web:integration_tests_linux_chrome
```
Comment (nturgut) [inline .cirrus.yml:98]: I removed this part from cirrus, since we are already deprecating cirrus: https://github.com/flutter/engine/pull/18035

File: .cirrus.yml [hunk 5]
```diff
@@ -181,10 +184,11 @@ task:
...
         cd $ENGINE_PATH/src/flutter/web_sdk/web_engine_tester
         $ENGINE_PATH/src/out/host_debug_unopt/dart-sdk/bin/pub get
-        cd $ENGINE_PATH/src/flutter/lib/web_ui
+        cd $ENGINE_PATH/src/flutter/tools/luci
         $ENGINE_PATH/src/out/host_debug_unopt/dart-sdk/bin/pub get
-        export FELT="$ENGINE_PATH/src/out/host_debug_unopt/dart-sdk/bin/dart dev/felt.dart"
-        $FELT test --browser=firefox
+        export LUCI="$ENGINE_PATH/src/out/host_debug_unopt/dart-sdk/bin/dart bin/luci.dart"
+        $LUCI run //lib/web_ui:unit_tests_linux_firefox
+
```
Comment (nturgut) [inline .cirrus.yml:191]: I'll remove this soon. I'm carrying it to LUCI: https://github.com/flutter/flutter/issues/55978

**Comment by yjbanov** [comment]
(moving this back to draft mode and discussion on the final design still continue)

**Comment by yjbanov** [comment]
The design has been substantially revised. This PR is no longer relevant.
